File: game/validation.py
from core.action import ActionType
from core.board_state import BoardState
from core.vector2 import Vector2


def validate_move(player, opponent, move):
    board_state = move.board_state
    action = move.action

    if action is None:
        return False

    if action.type == ActionType.MOVE:
        # check if move to a valid location
        valid_moves = player.get_valid_move_positions(board_state)
        if not any(action.new_pos.x == pos.x and action.new_pos.y == pos.y for pos in valid_moves):
            return False
    else:
        # player has enough walls
        if player.wall_count == 0:
            return False
        # wall is within bounds
        if not board_state.is_wall_index_in_bounds(action.block_pos):
            return False
        # wall is not on top of another wall
        if board_state.walls[action.block_pos.x, action.block_pos.y] != 0:
            return False
        # wall is not directly next to another wall of the same orientation
        shift_amt = Vector2(1, 0) if action.block_orientation == 2 else Vector2(0, 1)
        adjacent_point_1 = action.block_pos - shift_amt
        adjacent_point_2 = action.block_pos + shift_amt
        if board_state.is_wall_index_in_bounds(adjacent_point_1) and board_state.walls[adjacent_point_1.x, adjacent_point_1.y] == action.block_orientation:
            return False
        if board_state.is_wall_index_in_bounds(adjacent_point_2) and board_state.walls[adjacent_point_2.x, adjacent_point_2.y] == action.block_orientation:
            return False
        # player is not boxed in
        clone = board_state.clone()
        clone.walls[action.block_pos.x, action.block_pos.y] = action.block_orientation
        player_distance_matrix = clone.get_distance_matrix_from_row(player.goal_row)
        if player_distance_matrix[player.pos.x][player.pos.y] is None:
            return False
        opponent_distance_matrix = clone.get_distance_matrix_from_row(opponent.goal_row)
        if opponent_distance_matrix[opponent.pos.x][opponent.pos.y] is None:
            return False
    return True


def get_valid_move_positions(pos, board_state):
    directions = [Vector2(1, 0), Vector2(-1, 0), Vector2(0, 1), Vector2(0, -1)]
    for direction in directions:
        new_pos = pos + direction
        cell_is_valid = board_state.is_cell_index_in_bounds(new_pos)
        cell_is_blocked = cell_is_valid and board_state.is_path_blocked(pos, direction)
        cell_is_occupied = cell_is_valid and board_state.cells[new_pos.x, new_pos.y] != 0
        if cell_is_valid and not cell_is_blocked:
            if not cell_is_occupied:
                yield new_pos
            else:
                for jump_direction in directions:
                    jump_pos = new_pos + jump_direction
                    jump_cell_is_valid = board_state.is_cell_index_in_bounds(jump_pos)
                    jump_cell_is_blocked = jump_cell_is_valid and board_state.is_path_blocked(new_pos, jump_direction)
                    jump_cell_is_occupied = jump_cell_is_valid and board_state.cells[jump_pos.x, jump_pos.y] != 0
                    if jump_cell_is_valid and not jump_cell_is_blocked and not jump_cell_is_occupied:
                        yield jump_pos

File: game/src/validation.cc
#include "validation.h"
#include <algorithm>
#include <iostream>

using namespace std;

void PrintError(const string& message, const Action& action) {
    cout << "Invalid Action - " << message << " (" << action << ")" << endl;    
}

bool ValidateAction(const BoardState& board_state, const int player_index, const Action& action, const bool print_error) {
    if (action.GetType() == ActionType::MOVE) {
        // Check if move is to a valid location
        auto valid_moves = board_state.GetValidMoves(player_index);
        if (find(valid_moves.begin(), valid_moves.end(), action.GetMovePosition()) == valid_moves.end()) {
            if (print_error)
                PrintError("Move location invalid", action);
            return false;
        }
    }
    else {
        // Player has enough walls
        if (board_state.GetPlayerWallCount(player_index) == 0) {
            if (print_error)
                PrintError("Player has insufficient walls", action);
            return false;
        }
        // Wall is within bounds
        if (!BoardState::IsWallIndexInBounds(action.GetBlockPosition())) {
            if (print_error)
                PrintError("Wall out of bounds", action);
            return false;
        }
        // Wall is not on top of another wall
        if (board_state.GetWall(action.GetBlockPosition()) != 0) {
            if (print_error)
                PrintError("Fully overlapping walls", action);
            return false;
        }
        // Wall is not directly next to another wall of the same orientation
        const auto shift_amt = (action.GetBlockOrientation() == WallOrientation::HORIZONTAL) ? Vectori(1, 0) : Vectori(0, 1);
        const auto adjacent_point_1 = action.GetBlockPosition() - shift_amt;
        const auto adjacent_point_2 = action.GetBlockPosition() + shift_amt;
        if (BoardState::IsWallIndexInBounds(adjacent_point_1)
            && board_state.GetWall(adjacent_point_1) == action.GetBlockOrientation()) {
            if (print_error)
                PrintError("Partially overlapping walls", action);
            return false;
        }            
        if (BoardState::IsWallIndexInBounds(adjacent_point_2)
            && board_state.GetWall(adjacent_point_2) == action.GetBlockOrientation()) {
            if (print_error)
                PrintError("Partially overlapping walls", action);
            return false;
        }
        // Player is not boxed in
        const auto copy = BoardState(board_state, action, player_index);   
        if (IsPlayerTrapped(copy, player_index)) {
            if (print_error)
                PrintError("Player trapped", action);
            return false;
        }
        const auto opponent_index = 1 - player_index;
        if (IsPlayerTrapped(copy, opponent_index)) {
            if (print_error)
                PrintError("Opponent trapped", action);
            return false;
        }
    }
    return true;
}

bool IsValidWall(const BoardState& board_state, const Vectori position, const int orientation){
    if(board_state.GetWall(position) != 0)
        return false;
    const auto shift_amount = orientation == 1 ? Vectori(0, 1) : Vectori(1, 0);
    const auto point_a = position + shift_amount;
    if (BoardState::IsWallIndexInBounds(point_a) && board_state.GetWall(point_a) == orientation)
        return false;
    const auto point_b = position - shift_amount;
    if (BoardState::IsWallIndexInBounds(point_b) && board_state.GetWall(point_b) == orientation)
        return false;
    return true;
}

bool IsPlayerTrapped(const BoardState& board_state, const int player_index) {
    return board_state.GetPlayerDistance(player_index) == -1;
}

bool IsEitherPlayerTrapped(const BoardState& board_state) {
    return IsPlayerTrapped(board_state, 0) || IsPlayerTrapped(board_state, 1);
}